src/services/stringCompletions.ts

```diff
@@ -505,37 +522,44 @@ namespace ts.Completions.StringCompletions {
-            if (typeof typesVersions === "object") {
-                const versionResult = getPackageJsonTypesVersionsPaths(typesVersions);
-                const versionPaths = versionResult && versionResult.paths;
-                const rest = absolutePath.slice(ensureTrailingDirectorySeparator(baseDirectory).length);
-                if (versionPaths) {
-                    addCompletionEntriesFromPaths(result, rest, baseDirectory, extensions, versionPaths, host);
-                }
-            }
-        }
-
         return result;
     }
 
     function addCompletionEntriesFromPaths(result: NameAndKind[], fragment: string, baseDirectory: string, fileExtensions: readonly string[], paths: MapLike<string[]>, host: LanguageServiceHost) {
+        let pathResults: { results: NameAndKind[], matchedPattern: boolean }[] = [];
+        let matchedPathPrefixLength: number | undefined;
         for (const path in paths) {
             if (!hasProperty(paths, path)) continue;
             const patterns = paths[path];
             if (patterns) {
-                for (const { name, kind, extension } of getCompletionsForPathMapping(path, patterns, fragment, baseDirectory, fileExtensions, host)) {
-                    // Path mappings may provide a duplicate way to get to something we've already added, so don't add again.
-                    if (!result.some(entry => entry.name === name)) {
-                        result.push(nameAndKind(name, kind, extension));
+                const pathPattern = tryParsePattern(path);
+                if (!pathPattern) continue;
+                const isMatch = typeof pathPattern === "object" && isPatternMatch(pathPattern, fragment);
+                const isLongestMatch = isMatch && (matchedPathPrefixLength === undefined || pathPattern.prefix.length > matchedPathPrefixLength);
+                if (isLongestMatch) {
+                    // If this is a higher priority match than anything we've seen so far, previous results from matches are invalid, e.g.
+                    // {
+                    //   "bar/*": ["bar/*"], // <-- 1. We add 'bar', but 'bar/*' doesn't match yet.
+                    //   "*": ["dist/*"],   //  <-- 2. We match here. 'bar' is still ok because it didn't come from a match.
+                    //   "foo/*": ["foo/*"] //  <-- 3. We matched '*' earlier and added results from dist, but if 'foo/*' also matches,
+                    // }                               results in dist are actually not visible. 'bar' still stands because it didn't come from a match.
+                    matchedPathPrefixLength = pathPattern.prefix.length;
+                    pathResults = pathResults.filter(r => !r.matchedPattern);
+                }
+                if (typeof pathPattern === "string" || matchedPathPrefixLength === undefined || pathPattern.prefix.length >= matchedPathPrefixLength) {
+                    const pathResult: typeof pathResults[number] = { results: [], matchedPattern: isMatch };

```

Comment:
[Wow, you are really trying to mess with me today, huh?](https://github.com/microsoft/TypeScript/issues/19707)